resource_activity/models/resource_available.py

```diff
@@ -34,40 +39,38 @@ class ResourceAvailable(models.Model):
 
     @api.multi
     def action_reserve(self):
-        for resource_available in self.filtered(lambda record: record.state == "free"):
-            # todo pass registration / delegate to registration
-            allocation_ids = resource_available.resource_id.allocate_resource(
-                resource_available.registration_id.booking_type,
-                resource_available.registration_id.date_start,
-                resource_available.registration_id.date_end,
-                resource_available.registration_id.attendee_id,
-                resource_available.registration_id.location_id,
-                resource_available.registration_id.date_lock,
-            )
-            if allocation_ids:
-                allocations = self.env["resource.allocation"].browse(allocation_ids)
-                allocations.write(
-                    {"activity_registration_id": resource_available.registration_id.id}
-                )
-                resource_available.state = "selected"
-                # resource_available.registration_id.quantity_allocated += 1  # mark
-                resource_available.registration_id.state = (
-                    resource_available.registration_id.booking_type
-                )
-            else:
-                _logger.info(
-                    "no resource found for : " + str(resource_available.resource_id.ids)
+        # this code works because it is called from
+        # a list view on an activity form after button search_resource
+        # was called.
+        # In other cases, registration might not be set
+        for available_resource in self:
+            if not available_resource.registration_id:
+                raise ValueError(
+                    "Registration is not set on resource availability record %s."
+                    % available_resource.id
                 )
-            self.activity_id.registrations.action_refresh()
-        return True
+
+            available_resource.allocation_id = self.env["resource.allocation"].create(
+                {
+                    "resource_id": available_resource.resource_id.id,
+                    "partner_id": available_resource.registration_id.attendee_id.id,
+                    "date_start": available_resource.registration_id.date_start,
+                    "date_end": available_resource.registration_id.date_end,
+                    "date_lock": available_resource.registration_id.date_lock,
+                    "state": available_resource.registration_id.booking_type,
+                    "location": available_resource.registration_id.location_id.id,
+                    "activity_registration_id": available_resource.registration_id.id,
+                }
+            )
+            available_resource.state = "selected"
+            available_resource.activity_id.registrations.action_refresh()
```

Comment:
this should not be needed with correct dependencies on computed fields. not blocking as it was already there, and we’re not fixing all issues here.

Reply:
I agree.